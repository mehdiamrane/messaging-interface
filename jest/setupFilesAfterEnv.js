import '@testing-library/jest-dom/extend-expect';

jest.mock('next-i18next', () => ({
  useTranslation: () => ({ t: (key) => key }),
}));
