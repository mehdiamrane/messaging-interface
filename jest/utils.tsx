import React from 'react';
import { render, type RenderOptions, type RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import theme from 'src/styles/theme';

Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

const AllProviders = ({ children }: { children: React.ReactNode }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

interface Render {
  (ui: React.ReactElement, options?: RenderOptions): RenderResult;
}

const customRender: Render = (ui, options) =>
  render(ui, {
    wrapper: (props) => <AllProviders {...props} />,
    ...options,
  });

export * from '@testing-library/react';
export { customRender as render };
