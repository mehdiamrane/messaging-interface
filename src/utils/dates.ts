import formatDistanceToNow from 'date-fns/formatDistanceToNow';
import intlFormat from 'date-fns/intlFormat';
import { enUS, fr } from 'date-fns/locale';

const localeMapping = { 'en-US': enUS, fr };

export const formatRelativeDate = (date: Date, locale: string): string =>
  formatDistanceToNow(date, { locale: localeMapping[locale], addSuffix: true });

export const formatDate = (date: Date, locale: string): string =>
  intlFormat(
    date,
    {
      weekday: 'long',
      year: 'numeric',
      month: 'long',
      day: 'numeric',
      hour: 'numeric',
      minute: 'numeric',
    },
    {
      locale,
    },
  );

export const timestampToDate = (timestamp: number): Date => new Date(timestamp * 1000);
