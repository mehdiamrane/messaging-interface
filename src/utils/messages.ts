import type { Message } from 'src/types/message';
import type { ChatBubbleProps } from 'src/components/conversation/chatBubble/ChatBubble';

export const sortMessages = (messages: Message[]): Message[] =>
  (messages || []).sort((a, b) => a.timestamp - b.timestamp);

export const getMessagePosition = (messages: Message[], index: number): ChatBubbleProps['position'] => {
  const previous = messages[index - 1];
  const current = messages[index];
  const next = messages[index + 1];
  if (previous?.authorId !== current.authorId && next?.authorId !== current.authorId) {
    return 'single';
  }
  if (previous?.authorId !== current.authorId && next?.authorId === current.authorId) {
    return 'first';
  }
  if (previous?.authorId === current.authorId && next?.authorId !== current.authorId) {
    return 'last';
  }
  return 'middle';
};
