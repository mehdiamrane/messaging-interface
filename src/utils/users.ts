import type { User } from 'src/types/user';

// Default way to use a logged user
export const getLoggedUserId = (): User['id'] => 1;

export const isLoggedUser = (userId: User['id']): boolean => userId === getLoggedUserId();

export const getUserData = (userId: User['id'], users: User[]): User | undefined =>
  users.find(({ id }) => id === userId);

export const getAvatarFromNickname = (nickname: User['nickname']): string =>
  `https://ui-avatars.com/api/?bold=true&size=54&background=CBD5E0&rounded=true&name=${nickname}`;
