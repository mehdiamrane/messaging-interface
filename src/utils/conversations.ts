import type { Conversation } from 'src/types/conversation';
import type { BareUser } from 'src/types/user';
import { getLoggedUserId } from './users';

export const sortConversationsByLastActive = (conversations: Conversation[]): Conversation[] =>
  conversations.sort((a, b) => b.lastMessageTimestamp - a.lastMessageTimestamp);

export const getOtherUserFromConversation = (conversation: Conversation): BareUser =>
  conversation.senderId === getLoggedUserId()
    ? { id: conversation.recipientId, nickname: conversation.recipientNickname }
    : { id: conversation.senderId, nickname: conversation.senderNickname };
