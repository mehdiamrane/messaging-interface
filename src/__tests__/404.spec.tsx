import React from 'react';
import { render } from 'jest/utils';
import { cleanup } from '@testing-library/react';
import Page404 from 'src/pages/404';
import 'jest-styled-components';

const useRouter = jest.spyOn(require('next/router'), 'useRouter');
afterAll(cleanup);

describe('Page404', () => {
  it('should render correctly Page404', () => {
    useRouter.mockImplementation(() => ({
      locale: 'en-US',
    }));
    const { container } = render(<Page404 />);
    expect(container).toMatchSnapshot();
  });
});
