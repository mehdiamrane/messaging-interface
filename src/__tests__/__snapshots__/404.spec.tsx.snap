// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page404 should render correctly Page404 1`] = `
.c0 {
  display: initial;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
}

.c2 {
  background-color: rgba(0,0,0,0.04);
  border: 2px solid rgba(0,0,0,0.24);
  border-radius: 0.5rem;
  color: rgba(0,0,0,0.48);
  font-size: 1rem;
  font-weight: 500;
  line-height: 1.5;
  max-width: 24rem;
  padding: 3rem;
  width: 100%;
}

@media (min-width:48em) {
  .c0 {
    display: initial;
    -webkit-flex: 1;
    -ms-flex: 1;
    flex: 1;
    width: 100%;
  }
}

<div>
  <main
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        error.404.text
      </div>
    </div>
  </main>
</div>
`;
